### ISSUE TITLE
`option_if_let_else` misses lint on `mut` version of function

### ISSUE THREAD

**Issue by OliveIsAWord** [rust-lang/rust-clippy#8967]
### Summary

The nursery lint `option_if_let_else` will correctly catch a function working with `Option<&T>` but will not catch a nearly identical function working with `Option<&mut T>`.

### Lint Name

option_if_let_else

### Reproducer

I tried this code:

```rust
#![warn(clippy::option_if_let_else)]

pub struct HashStack {
    scope: Vec<i32>,
    parent: Option<Box<Self>>,
}

impl HashStack {
    pub fn lookup(&self, index: usize) -> Option<&i32> {
        if let Some(val) = self.scope.get(index) {
            Some(val)
        } else {
            self.parent.as_ref().and_then(|p| p.lookup(index))
        }
    }

    pub fn lookup_mut(&mut self, index: usize) -> Option<&mut i32> {
        if let Some(val) = self.scope.get_mut(index) {
            Some(val)
        } else {
            self.parent.as_mut().and_then(|p| p.lookup_mut(index))
        }
    }
}
```

The lint warns on the first method, and suggests this fix:
```rust
self.scope.get(index).map_or_else(
    || self.parent.as_ref().and_then(|p| p.lookup(index)),
    |val| Some(val),
)
```
However, it does not warn on the second method. I expected it to warn and suggest this similar fix:
```rust
self.scope.get_mut(index).map_or_else(
    || self.parent.as_mut().and_then(|p| p.lookup_mut(index)),
    |val| Some(val),
)
```

### Version

```text
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-pc-windows-msvc
release: 1.61.0
LLVM version: 14.0.0
```

**Comment by ada4a** [comment]
@blyxyas I'd like to work on this, but it'll probably count as a feature addition which we currently avoid, right?
